distsql: introduce SpanResolver

andreimatei opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10575)
<!-- Reviewable:end -->

andreimatei committed:
distsql: introduce SpanResolver

SpanResolver will be used by physical planning for figuring out the
mapping between key spans and kv ranges.

andreimatei committed:
server: move LookupRange and SplitRange to TestServer

... from TestCluster, so that they can be used by tests that only need a
single node. Also change their results from pointers to vals.

andreimatei commented:
still working on more tests

---

Review status: 0 of 18 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWA8L9MiH0EKaYT9aWM:b-y3eyfp)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks a lot for getting this done!! WOOOHOO

Most comments I have are from the perspective of a prospective user. Someone more familiar with the lower layers that we're using should also take a look

---

Review status: 0 of 18 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAgqXXNyF8Rmx469Ln:-KWAgqXXNyF8Rmx469Lo:blh9gni) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L74)):*
> ```Go
> const (
> 	// Ascending means Key is inclusive and EndKey is exclusive.
> 	Ascending ScanType = iota
> ```

Can't we use an existing type for this (like encoding.Direction)? Or if the direction is actually not important and we only care about which end is inclusive, maybe name them accordingly (eg StartInclusive, EndInclusive)

---

*[pkg/sql/distsql/span_resolver.go, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAh8Y6lDVSC2Cq1SHI:-KWAh8Y6lDVSC2Cq1SHJ:b-qvfuaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L93)):*
> ```Go
> 	case BinPackingLHGuessingPolicy:
> 		oracle = &binPackingOracle{
> 			// This number is based on nothing.
> ```

duplicate comment

---

*[pkg/sql/distsql/span_resolver.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAgzr8pPmBs8tWLDgF:-KWAgzr8pPmBs8tWLDgG:b-lbsygz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L108)):*
> ```Go
> }
> 
> func (sr *SpanResolver) resolveSpan(
> ```

could use  a comment

---

*[pkg/sql/distsql/span_resolver.go, line 132 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAmDEjx2-vnM6rVWS3:-KWAmDEjx2-vnM6rVWS4:b-li3a46) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L132)):*
> ```Go
> 
> 	var leaseHolders []kv.ReplicaInfo
> 	ri := kv.NewRangeIterator(sr.distSender, scanType == Descending)
> ```

can/should we reuse the iterator? it may make sense to continue seeking from where we left off

---

*[pkg/sql/distsql/span_resolver.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAmn7Kc9smUQtu7O0M:-KWAmn7Kc9smUQtu7O0N:by3vb3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L160)):*
> ```Go
> 		}
> 	}
> 	return leaseHolders, nil
> ```

If `ri.Seek()` runs into an error, shouldn't we return `ri.Error()`?

---

*[pkg/sql/distsql/span_resolver.go, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAkszW6cOH8Xnn4yzb:-KWAkszW6cOH8Xnn4yzc:btpiy79) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L166)):*
> ```Go
> // spans are distributed across kv ranges.
> // For each input span it returns an array of results containing information
> // about all the ranges spanned by the respective span. Each range is
> ```

What is not ideal here is that if we have a lot of spans within a replica, we have a separate copy of the same ReplicaInfo for each span. I'm thinking that perhaps in the future we will want have a route-by-range that uses this API with a lot of small spans (rows).

The calling code will have to go through all the spans and do some work to trim the spans to the range boundaries, so it could just as easily handle a `[]kv.ReplicaInfo` which just contains all ranges touched by the spans (in order).

---

*[pkg/sql/distsql/span_resolver.go, line 169 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAoQ8Jc4LHAlrcxr53:-KWAoQ8Jc4LHAlrcxr54:b-mev7pk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L169)):*
> ```Go
> // represented by a RangeInfo, which identifies the replica that is believed
> // to be the current lease holder of the range. Information about connecting to
> // the node owning that replica is also included.
> ```

Is it guaranteed that the ranges for each span cover the span without gaps? (is the end key of each range the same with the start key of the next?)

---

*[pkg/sql/distsql/span_resolver.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAi3UhvFOAYzBwBIFH:-KWAi3UhvFOAYzBwBIFI:b-p136ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L177)):*
> ```Go
> // spans but not necessarily the EndKey (so StartKey is inclusive, EndKey is
> // exclusive). If it's Descending then the opposite holds. Regardless of
> // scanType, the input spans still need to be sorted ascendingly.
> ```

should we verify the sort order and error out or panic, in case the higher layers are doing something wrong?

---

*[pkg/sql/distsql/span_resolver.go, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAiJXhn8Iznh8du5mk:-KWAiJXhn8Iznh8du5ml:bizaom) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L206)):*
> ```Go
> 	} else {
> 		// Iterate backwards so that prefetching helps us across spans.
> 		for i := len(spans) - 1; i >= 0; i-- {
> ```

we could have a single loop `for i := start; i != end; i += step {}` and initialize `start`, `end`, `step` accordingly

---

*[pkg/sql/distsql/span_resolver.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAn0uThkha_CQQ3wd0:-KWAn0uThkha_CQQ3wd1:brd8foo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L208)):*
> ```Go
> 		for i := len(spans) - 1; i >= 0; i-- {
> 			var err error
> 			leaseHolders[i], err = sr.resolveSpan(ctx, spans[i], scanType, rangesPerNode)
> ```

We could check if the span is already completely inside the last `kv.ReplicaInfo` and skip the call in that case. This ties in with the comment above to maybe just return a single list of replica infos.

---

*[pkg/sql/distsql/span_resolver.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAj0eFddUybitcuxet:-KWAj0eFddUybitcuxeu:bhcvcom) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L233)):*
> ```Go
> 	// about any of the nodes that might be tried.
> 	GuessLeaseHolder(
> 		desc roachpb.RangeDescriptor, rangesPerLeaseHolder map[roachpb.NodeID]uint,
> ```

Maybe rangesPerLeaseHolder belongs in a separate struct where we explain that it contains stuff that oracle implementations can use to make their decisions. It would make it cleaner and easier to add more stuff.

Also, why `uint`? I'd prefer `int` if you don't have a strong reason (unexpected negative values are usually easier to debug vs underflows).

---

*[pkg/sql/distsql/span_resolver.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAjxPApGodwgG3tD0L:-KWAjxPApGodwgG3tD0M:b-tcnkom) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L298)):*
> ```Go
> 	// full. Use the least-loaded one.
> 	return replicas[leastLoadedIdx], nil
> ```

[nit] extra line

---

*[pkg/sql/distsql/span_resolver_test.go, line 199 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWApSJa2-j7LMJdwdpK:-KWApSJa2-j7LMJdwdpL:b-xuvddw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver_test.go#L199)):*
> ```Go
> // `CREATE TABLE test (k INT PRIMARY KEY)` at row with value pk (the row will be
> // the first on the right of the split).
> func splitRangeAtKey(
> ```

We already have SQL for this, why not use that?

---

*[pkg/sql/distsql/span_resolver_test.go, line 220 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAqwA6m4MJz0kRwOQL:-KWAqwA7rIcvNVVDp1Se:b-5nwsap) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver_test.go#L220)):*
> ```Go
> 	pik := primaryIndexKey(tableDesc, parser.NewDInt(parser.DInt(pk)))
> 	startKey := keys.MakeFamilyKey(pik, uint32(tableDesc.Families[0].ID))
> 	leftRange, rightRange, err := ts.SplitRange(startKey)
> ```

note that in the implementation of SPLIT AT, we had to tolerate occasional `conflict updating range descriptors` errors.. If you keep this code please run the tests under stress to make sure they're not flaky. Come to think of it, I believe @tamird ensured that it's done for you now :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWAq-Us0_-dkW4_Jobi:bp1alxf)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Transformed the code into an iterator, see if you like it. I think it works more naturally with the lower layer, that's also an iterator, and it makes more clear the significance of that "scan direction" argument, and also makes it more natural for a client to mix directions between spans... Plus I also think we'll need to use an iteration model for `SELECT * FROM t LIMIT 1` from day one... And probably also that query + a filter.

---

Review status: 0 of 19 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAgqXXNyF8Rmx469Ln:-KWE9QetQCZFEzgf2qaD:b-qmsb08) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L74)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Can't we use an existing type for this (like encoding.Direction)? Or if the direction is actually not important and we only care about which end is inclusive, maybe name them accordingly (eg StartInclusive, EndInclusive)
</details>
I've considered various alternatives... 
- `encoding.Direction` is a much lower-level concept.
- the direction is actually relevant to the implementation (as an optimization), and might also become important to the interface too if we move this Resolver to an iterator model in the future.
- the "scan direction" concept with its associated semantics about which end is inclusive is very pervasive in our code base, for better or worse. Shows up all over the place, in different APIs and caches. Everybody and their dog takes a `reverse bool`, which nobody explains. I couldn't bare that bool any more, with all the comments at call sites about what it is. The enum is more evident and also gives you a place to hang documentation off of.
I think it'd be nice if some of the lower levels moved to more explicit arguments, but at this level here I think it's fine.

---

*[pkg/sql/distsql/span_resolver.go, line 166 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAkszW6cOH8Xnn4yzb:-KWEcyvlyfYTQt0NJvVC:b-ep44k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L166)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What is not ideal here is that if we have a lot of spans within a replica, we have a separate copy of the same ReplicaInfo for each span. I'm thinking that perhaps in the future we will want have a route-by-range that uses this API with a lot of small spans (rows).
> 
> The calling code will have to go through all the spans and do some work to trim the spans to the range boundaries, so it could just as easily handle a `[]kv.ReplicaInfo` which just contains all ranges touched by the spans (in order).
</details>
the iterator now makes this optimization

---

*[pkg/sql/distsql/span_resolver.go, line 169 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAoQ8Jc4LHAlrcxr53:-KWEcB07w_W2VMBT1MFr:b-dmhzfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L169)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Is it guaranteed that the ranges for each span cover the span without gaps? (is the end key of each range the same with the start key of the next?)
</details>
Yes.
The lower levels take care of this, in the face of inconsistent range descriptor reads racing with splits.

---

*[pkg/sql/distsql/span_resolver.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAj0eFddUybitcuxet:-KWFw_W3bGbsxExx9qbb:b5130sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L233)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Maybe rangesPerLeaseHolder belongs in a separate struct where we explain that it contains stuff that oracle implementations can use to make their decisions. It would make it cleaner and easier to add more stuff.
> 
> Also, why `uint`? I'd prefer `int` if you don't have a strong reason (unexpected negative values are usually easier to debug vs underflows).
</details>
Done. I agree with the separate struct. Initially I wanted to make it a black box implemented by each oracle, but the code above the iterator needs some access to updating it because, if a lease holder is known, the oracle is not involved, so... meh.

About the `uint`... would you then say that `uint` is pretty much never to be used (unless you need the 1 bit of extra range)? But done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWFx9g0KLsv-VOoHlSG:b-slis4l)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 
But also get a review from someone with deeper `kv` knowledge, maybe @petermattis

---

Review status: 0 of 19 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAgqXXNyF8Rmx469Ln:-KWG8tH1EbD20qtGlozk:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L74)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I've considered various alternatives... 
> - `encoding.Direction` is a much lower-level concept.
> - the direction is actually relevant to the implementation (as an optimization), and might also become important to the interface too if we move this Resolver to an iterator model in the future.
> - the "scan direction" concept with its associated semantics about which end is inclusive is very pervasive in our code base, for better or worse. Shows up all over the place, in different APIs and caches. Everybody and their dog takes a `reverse bool`, which nobody explains. I couldn't bare that bool any more, with all the comments at call sites about what it is. The enum is more evident and also gives you a place to hang documentation off of.
> I think it'd be nice if some of the lower levels moved to more explicit arguments, but at this level here I think it's fine.
</details>
:thumbsup: 

---

*[pkg/sql/distsql/span_resolver.go, line 169 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAoQ8Jc4LHAlrcxr53:-KWG95xEBSW1SjO7obC3:bb76rtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L169)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Yes.
> The lower levels take care of this, in the face of inconsistent range descriptor reads racing with splits.
</details>
Woot

---

*[pkg/sql/distsql/span_resolver.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAj0eFddUybitcuxet:-KWGDv3n_mVI1p0U_znx:bm7vkxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L233)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done. I agree with the separate struct. Initially I wanted to make it a black box implemented by each oracle, but the code above the iterator needs some access to updating it because, if a lease holder is known, the oracle is not involved, so... meh.
> 
> About the `uint`... would you then say that `uint` is pretty much never to be used (unless you need the 1 bit of extra range)? But done.
</details>
In general, yeah.. And seems that Go went that way with e.g. `len()` and stuff. Just a suggestion though, feel free to ignore.

---

*[pkg/sql/distsql/span_resolver.go, line 171 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWG9iVZs502f-K4sy25:-KWG9iVZs502f-K4sy26:bmfb7yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5baa7fe1db82f137638d410185a9e193cabb7b/pkg/sql/distsql/span_resolver.go#L171)):*
> ```Go
> 
> // Seek positions the iterator on the start of a span.
> // After calling this, ReplicaInfo() will return information about the first
> ```

Maybe "the range containing the start of the span" is a bit more explicit

---

*[pkg/sql/distsql/span_resolver.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWGAZtoF8suF50FRqtA:-KWGAZtpb9tNvNCaUlGR:brc61cm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5baa7fe1db82f137638d410185a9e193cabb7b/pkg/sql/distsql/span_resolver.go#L173)):*
> ```Go
> // After calling this, ReplicaInfo() will return information about the first
> // range of the span. NeedAnother() will return true until the iterator is
> // positioned on the end of the span.
> ```

on or after the end of the span?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWGF5LTZmKNK8bybTM-:by988u)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
A few changes:
- added a `Desc()` method for getting the current `RangeDesc`
- removed all language about the scan direction affecting the inclusive nature of `EndKey`

@raduberinde, I've kept the resolving of nodes to addresses inside the iterator, even though it could be done outside and only once per node. This is because, when an oracle is involved, we need to resolve through gossip to get the node attributes. And also because we avoid returning replicas on nodes that are not available in gossip.
The sanity of the latter point is dubious, but it's also what the DistSender is doing.

In any case, hopefully this isn't expensive. If it is, the DistSender has the same problem and we can likely improve the gossip lookup process.

---

Review status: 0 of 21 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 93 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAh8Y6lDVSC2Cq1SHI:-KWf7kt6m0So1tw6OdZN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L93)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> duplicate comment
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAgzr8pPmBs8tWLDgF:-KWf8YJOIIcLYoa8UEzF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> could use  a comment
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 132 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAmDEjx2-vnM6rVWS3:-KWf99_D2wAH11IjD93c:b-qb99k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> can/should we reuse the iterator? it may make sense to continue seeking from where we left off
</details>
sorry, I don't follow. If this condition passes, we will reuse the `it.it`.

---

*[pkg/sql/distsql/span_resolver.go, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAi3UhvFOAYzBwBIFH:-KWf9LjJFuGaNnHAq0G8:bvkrcgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L177)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> should we verify the sort order and error out or panic, in case the higher layers are doing something wrong?
</details>
no longer applicable

---

*[pkg/sql/distsql/span_resolver.go, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAiJXhn8Iznh8du5mk:-KWf9QLNB0UtWWb8c-WE:bvkrcgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L206)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> we could have a single loop `for i := start; i != end; i += step {}` and initialize `start`, `end`, `step` accordingly
</details>
no longer applicable

---

*[pkg/sql/distsql/span_resolver.go, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAn0uThkha_CQQ3wd0:-KWf9Sfz5zor2InE1Evq:bvkrcgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L208)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> We could check if the span is already completely inside the last `kv.ReplicaInfo` and skip the call in that case. This ties in with the comment above to maybe just return a single list of replica infos.
</details>
no longer applicable

---

*[pkg/sql/distsql/span_resolver.go, line 233 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAj0eFddUybitcuxet:-KWf9VvQbWCNrcA0DFHe:b-bpet43) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L233)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> In general, yeah.. And seems that Go went that way with e.g. `len()` and stuff. Just a suggestion though, feel free to ignore.
</details>
yeah, I don't disagree.

---

*[pkg/sql/distsql/span_resolver.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAjxPApGodwgG3tD0L:-KWf9aYWRnX_sfBqelzC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L298)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] extra line
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 171 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWG9iVZs502f-K4sy25:-KWf8sRA4O78jtZR9xP_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5baa7fe1db82f137638d410185a9e193cabb7b/pkg/sql/distsql/span_resolver.go#L171)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Maybe "the range containing the start of the span" is a bit more explicit
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWGAZtoF8suF50FRqtA:-KWf908Ua1T6C6485oy5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd5baa7fe1db82f137638d410185a9e193cabb7b/pkg/sql/distsql/span_resolver.go#L173)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> on or after the end of the span?
</details>
Done.

---

*[pkg/sql/distsql/span_resolver_test.go, line 199 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWApSJa2-j7LMJdwdpK:-KWivwZNRRirnN4lgsRo:bskt1i5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver_test.go#L199)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> We already have SQL for this, why not use that?
</details>
I might use it... Let me play around.

---

*[pkg/sql/distsql/span_resolver_test.go, line 220 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAqwA6m4MJz0kRwOQL:-KWisOpQjX8iWmsMyqzw:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver_test.go#L220)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> note that in the implementation of SPLIT AT, we had to tolerate occasional `conflict updating range descriptors` errors.. If you keep this code please run the tests under stress to make sure they're not flaky. Come to think of it, I believe @tamird ensured that it's done for you now :)
</details>
ack

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWiwFid455Fu64_IEBR:bj6gspv)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 21 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 950 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWj43mzyuL1rzyTKtgN:-KWj43mzyuL1rzyTKtgO:b-hbrjoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/index_selection.go#L950)):*
> ```Go
> 
> // Encodes datum at the end of key, using direction `dir` for the encoding.
> // It takes in an inclusive key and return an inclusive key if
> ```

returns

---

*[pkg/sql/distsql/span_resolver_test.go, line 450 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWj6Kju0bpnyVr-fty_:-KWj6Kju0bpnyVr-ftya:bx79h5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver_test.go#L450)):*
> ```Go
> 		res = append(res, repls)
> 	}
> 	return res, nil
> ```

If we exit the loop because `it.Valid()` is false, we should return `it.Error()` no? Maybe doing `if !it.Valid() { return nil. it.Error()` in the loop makes more sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWj6f0MNL3gmv8iGHO_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 9 files at r1, 3 of 13 files at r4, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 35 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlR1Lc4joYASckIuyg:-KWlR1Ld-888O2-BDnIx:bqsndvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L35)):*
> ```Go
> // When guessing lease holders, we try to guess the same node for all the ranges
> // applicable, until we hit this limit. The rationale is that maybe a bunch of
> // those ranges don't have an active lease, so our guess is going to be
> ```

Once we have epoch-based leases, it will be rare for a range to not have an active lease. 

---

*[pkg/sql/distsql/span_resolver.go, line 82 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlPPcNtFPhiJxhy5iO:-KWlPPcNtFPhiJxhy5iP:bx2twt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L82)):*
> ```Go
> 	// preference to the current node and others "close" to it when guessing lease
> 	// holders.
> 	// TODO(andrei): can the descriptor change at runtime?
> ```

Not currently, and there are other places where we store copies of the NodeDescriptor so we'd have to clean all of those up before we could allow runtime changes to the descriptor. Unlikely to happen any time soon.

---

*[pkg/sql/distsql/span_resolver.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlQHOfkmfUCplPSnhx:-KWlQHOfkmfUCplPSnhy:b-is6h26) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L304)):*
> ```Go
> 	// A RangeUnavailableError can be returned if there's no information in gossip
> 	// about any of the nodes that might be tried.
> 	GuessLeaseHolder(
> ```

This method seems misnamed. These aren't really trying to guess where the lease holder *is*, they're choosing where we would prefer the lease go if it's free. 

---

*[pkg/sql/sqlbase/errors.go, line 427 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMzO9Wcd192Lv-tgX:-KWlMzO9Wcd192Lv-tgY:b-gb7ckb) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/sqlbase/errors.go#L427)):*
> ```Go
> }
> 
> // ErrRangeUnavailable represents a missing database error.
> ```

s/database/range/

---

*[pkg/sql/sqlbase/errors.go, line 435 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlN9oV-RL59C1g7WfP:-KWlN9oV-RL59C1g7WfQ:bhl966v) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/sqlbase/errors.go#L435)):*
> ```Go
> }
> 
> // NewRangeUnavailableError creates a new ErrRangeUnavailable.
> ```

I know you're just following the pattern in this file, but it's unusual for the constructor name and the type name to differ. Most of our error types use the suffix `Error`. (why have I mentioned this in three PRs in the same day after never being an issue before?)

---

*[pkg/testutils/testcluster/testcluster.go, line 248 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMDgA1bSx7BDtvphI:-KWlMDgBg0Ljb9IGTzjI:b-hlkaku) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/testutils/testcluster/testcluster.go#L248)):*
> ```Go
> 	splitKey roachpb.Key,
> ) (roachpb.RangeDescriptor, roachpb.RangeDescriptor, error) {
> 	return tc.Servers[0].SplitRange(splitKey)
> ```

Merge conflict warning: I think @vivekmenezes will need to change this method in #10728.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWlRUmldXohE6WGk_l5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Unfortunately the tests were flaky because of #10751, whose investigation showed a number of weird things in how we manage the `RangeCache`; I've had to add a sleep in the test :) to make the test not flake. I think I'll merge like this and will work on fixing the various caching problems.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/index_selection.go, line 950 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWj43mzyuL1rzyTKtgN:-KWtdDL9Wo3ozInAWOGD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/index_selection.go#L950)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> returns
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWAmn7Kc9smUQtu7O0M:-KWtToTTDTJZ1gTPrfrg:bvkrcgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a79c535b387640fef11b38910f5cfa4a2b6c3558/pkg/sql/distsql/span_resolver.go#L160)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> If `ri.Seek()` runs into an error, shouldn't we return `ri.Error()`?
</details>
no longer applicable

---

*[pkg/sql/distsql/span_resolver.go, line 35 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlR1Lc4joYASckIuyg:-KWtVeet9HKgRhKA5Ph0:bxc5p9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L35)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Once we have epoch-based leases, it will be rare for a range to not have an active lease. 
</details>
yeah, I'm aware of this impedance mismatch. However I don't know what to do about it just now.

My plan is to move the whole planning process for large queries to an iterative model, where we plan the first ranges we want to read, then plan some more as we go. At that point, this `SpanResolverIterator` would resolve ranges in advance, diminishing the role of "guessing".
Separately, for small queries, I plan to do something so that information about bad planning flows back to the gateway node, updating its caches. We have that today with the DistSender, but we're losing it with DistSQL. This will further diminish the role of guessing for common queries.
And, of course, separate for all of these, we need some new plan for having data be located in accordance with access patterns. Not sure what to do here; generally it's a hard optimization problem. It'd probably be easy to get back the crude what we had today and are losing with epoch-based leases - we could add a queue that releases the leases for ranges that haven't been accessed in a while (I think it'd be easy to adapt the lease transfer code to work for "releasing" a lease too). But that's not actually what we want.

---

*[pkg/sql/distsql/span_resolver.go, line 82 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlPPcNtFPhiJxhy5iO:-KWtY5WBVeU_UCh7xEy1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L82)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Not currently, and there are other places where we store copies of the NodeDescriptor so we'd have to clean all of those up before we could allow runtime changes to the descriptor. Unlikely to happen any time soon.
</details>
Done.

---

*[pkg/sql/distsql/span_resolver.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlQHOfkmfUCplPSnhx:-KWtUCQGf9PB07Rs6Y1h:biyyhq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L304)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This method seems misnamed. These aren't really trying to guess where the lease holder *is*, they're choosing where we would prefer the lease go if it's free. 
</details>
well, the method does both. I'll take another name suggestion.

---

*[pkg/sql/distsql/span_resolver_test.go, line 450 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWj6Kju0bpnyVr-fty_:-KWtUr2gP0Sti8iNb07L:b3l85td) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver_test.go#L450)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> If we exit the loop because `it.Valid()` is false, we should return `it.Error()` no? Maybe doing `if !it.Valid() { return nil. it.Error()` in the loop makes more sense.
</details>
Done. good catch

---

*[pkg/sql/sqlbase/errors.go, line 427 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMzO9Wcd192Lv-tgX:-KWtV8AywyFBiyImfLYh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/sqlbase/errors.go#L427)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/database/range/
</details>
Done.

---

*[pkg/sql/sqlbase/errors.go, line 435 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlN9oV-RL59C1g7WfP:-KWtVPE9uEWlasfWRcMX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/sqlbase/errors.go#L435)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I know you're just following the pattern in this file, but it's unusual for the constructor name and the type name to differ. Most of our error types use the suffix `Error`. (why have I mentioned this in three PRs in the same day after never being an issue before?)
</details>
Done.

---

*[pkg/testutils/testcluster/testcluster.go, line 248 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMDgA1bSx7BDtvphI:-KWtZ-4ldZHv_jodqA8a:b3uw7gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/testutils/testcluster/testcluster.go#L248)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Merge conflict warning: I think @vivekmenezes will need to change this method in #10728.
</details>
I'm aware of that change, but I'm not sure what change to this `TestServer` method you have in mind exactly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-:-KWtdf9dvjUbdsRB_kOI:b6fkph5)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 4 of 21 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql/span_resolver.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlQHOfkmfUCplPSnhx:-KWvN1Y4AHSDenI4auWo:bfnxwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L304)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well, the method does both. I'll take another name suggestion.
</details>
Where is it guessing? I'd suggest something like ChoosePreferredLeaseHolder for the name. 

---

*[pkg/sql/distsql/span_resolver_test.go, line 220 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWvPgfutbDOkkDSWxpU:-KWvPgfutbDOkkDSWxpV:bbtjzdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4979dfc1a3ac1fef2bebce90333897df11aaf00/pkg/sql/distsql/span_resolver_test.go#L220)):*
> ```Go
> 		Knobs: base.TestingKnobs{
> 			Store: &storage.StoreTestingKnobs{
> 				// Disable the split queue, as it can rage with the splitting done
> ```

s/rage/race/

---

*[pkg/testutils/testcluster/testcluster.go, line 248 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMDgA1bSx7BDtvphI:-KWvQNi-HdcPNmY0_IOW:b25uj42) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/testutils/testcluster/testcluster.go#L248)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'm aware of that change, but I'm not sure what change to this `TestServer` method you have in mind exactly.
</details>
I'm arguing there that the callers of AdminSplit should be responsible for retrying instead of the lower levels. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 20 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsql/span_resolver.go, line 304 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlQHOfkmfUCplPSnhx:-KX8DEH-pyB4TselrmH-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/sql/distsql/span_resolver.go#L304)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Where is it guessing? I'd suggest something like ChoosePreferredLeaseHolder for the name. 
</details>
Done.

---

*[pkg/sql/distsql/span_resolver_test.go, line 220 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWvPgfutbDOkkDSWxpU:-KX8DZjDs_1IZwIzuVOl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4979dfc1a3ac1fef2bebce90333897df11aaf00/pkg/sql/distsql/span_resolver_test.go#L220)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/rage/race/
</details>
Done.

---

*[pkg/testutils/testcluster/testcluster.go, line 248 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575#-KWlMDgA1bSx7BDtvphI:-KX8FiWdFDL--nmBTw9k:bcgci8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/107b0b700175febe0911886cd1c2ff67af7e95a9/pkg/testutils/testcluster/testcluster.go#L248)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I'm arguing there that the callers of AdminSplit should be responsible for retrying instead of the lower levels. 
</details>
seems like you're losing that argument :)
In any case, if anything, `TestServer.SplitRange` will need to change, not this `TestCluster.SplitRange`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10575)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
🎉
